File: mobile/e_platform_mobile/lib/Service/progress/progress_service.dart
import '../../core/api/api_client.dart';
import '../../core/api/api_endpoints.dart';

/// Progress Service
/// Handles all progress-related API calls
class ProgressService {
  final ApiClient _apiClient = ApiClient();

  /// Get progress by enrollment
  Future<Map<String, dynamic>> getProgressByEnrollment(
    String enrollmentId,
  ) async {
    final response = await _apiClient.get(
      ApiEndpoints.progressByEnrollment(enrollmentId),
    );
    return response.data;
  }

  /// Update progress
  Future<Map<String, dynamic>> updateProgress(
    Map<String, dynamic> data,
  ) async {
    final response = await _apiClient.post(
      ApiEndpoints.progressUpdate,
      data: data,
    );
    return response.data;
  }

  /// Mark lesson as completed
  Future<Map<String, dynamic>> markLessonCompleted(
    String enrollmentId,
    String lessonId,
  ) async {
    final response = await _apiClient.post(
      ApiEndpoints.progressUpdate,
      data: {
        'enrollmentId': enrollmentId,
        'lessonId': lessonId,
        'completed': true,
      },
    );
    return response.data;
  }

  /// Get overall progress
  Future<Map<String, dynamic>> getOverallProgress() async {
    final response = await _apiClient.get(
      ApiEndpoints.progress,
    );
    return response.data;
  }

  /// Get progress stats
  Future<Map<String, dynamic>> getProgressStats(
    String enrollmentId,
  ) async {
    final response = await _apiClient.get(
      '${ApiEndpoints.progressByEnrollment(enrollmentId)}/stats',
    );
    return response.data;
  }

  // ========== ORCHESTRATOR METHODS (Item Completion) ==========

  /// Mark item as completed (ORCHESTRATOR)
  /// Marks specific item (lesson/quiz) as completed
  /// Cross-module: Progress + Enrollment modules
  Future<Map<String, dynamic>> markItemCompleted(
    String progressId,
    String itemId,
  ) async {
    final response = await _apiClient.post(
      ApiEndpoints.progressCompleteItem(progressId),
      data: {'itemId': itemId},
    );
    return response.data;
  }

  /// Get progress by ID (DIRECT)
  /// Single module operation
  Future<Map<String, dynamic>> getProgressById(String progressId) async {
    final response = await _apiClient.get(
      ApiEndpoints.progressById(progressId),
    );
    return response.data;
  }

  /// Get my progress (DIRECT)
  /// Single module operation - gets all progress for current user
  Future<List<Map<String, dynamic>>> getMyProgress() async {
    final response = await _apiClient.get(
      ApiEndpoints.myProgress,
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Get progress by course (DIRECT)
  /// Single module operation - gets progress summary for specific course
  Future<Map<String, dynamic>> getProgressByCourse(String courseId) async {
    final response = await _apiClient.get(
      ApiEndpoints.progressByCourse(courseId),
    );
    return response.data;
  }

  /// Get or create progress (DIRECT)
  /// Initializes or retrieves progress for a module
  Future<Map<String, dynamic>> getOrCreateProgress(
    Map<String, dynamic> data,
  ) async {
    final response = await _apiClient.post(
      ApiEndpoints.progress,
      data: data,
    );
    return response.data;
  }

  /// Get user course progress summary (DIRECT)
  /// Returns summary of progress for a specific course
  Future<Map<String, dynamic>> getUserCourseProgressSummary(
    String courseId,
  ) async {
    final response = await _apiClient.get(
      ApiEndpoints.progressByCourse(courseId),
    );
    return response.data;
  }

  /// Update progress by ID (DIRECT)
  /// Updates specific progress record
  Future<Map<String, dynamic>> updateProgressById(
    String progressId,
    Map<String, dynamic> updateData,
  ) async {
    final response = await _apiClient.patch(
      ApiEndpoints.progressById(progressId),
      data: updateData,
    );
    return response.data;
  }

  /// Get progress by module (DIRECT)
  /// Instructor view - get progress for specific module
  Future<List<Map<String, dynamic>>> getProgressByModule(
    String moduleType,
    String moduleId,
  ) async {
    final response = await _apiClient.get(
      '${ApiEndpoints.progress}/module/$moduleType/$moduleId',
    );
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Delete progress (admin only)
  Future<void> deleteProgress(String progressId) async {
    await _apiClient.delete(ApiEndpoints.progressById(progressId));
  }
}

File: web/e-platform-web/src/services/ProgressService.js
const API_URL = process.env.REACT_APP_BFF_API_URL || 'https://tunsiaed.onrender.com';
const getAuthHeaders = (token) => ({
  'Authorization': `Bearer ${token}`,
  'Content-Type': 'application/json',
});

/**
 * Progress Service - Handles all progress tracking operations
 * Maps to /api/v1/progress endpoints in the backend
 */
class ProgressService {
  /**
   * Get or create progress for a module
   * @param {string} token - Authentication token
   * @param {Object} data - { enrollmentId, moduleType, moduleId, totalItems }
   * @returns {Promise<Object>} Progress data
   */
  static async getOrCreateProgress(token, data) {
    const res = await fetch(`${API_URL}/api/v1/progress`, {
      method: 'POST',
      headers: getAuthHeaders(token),
      body: JSON.stringify(data),
    });
    if (!res.ok) throw new Error('Failed to get or create progress');
    return res.json();
  }

  /**
   * Get all progress for the authenticated user
   * @param {string} token - Authentication token
   * @returns {Promise<Object>} { progress: Array } - List of progress records
   */
  static async getMyProgress(token) {
    const res = await fetch(`${API_URL}/api/v1/progress/my-progress`, {
      headers: getAuthHeaders(token),
    });
    if (!res.ok) throw new Error('Failed to fetch progress');
    return res.json();
  }

  /**
   * Get progress by ID
   * @param {string} token - Authentication token
   * @param {string} progressId - The progress ID
   * @returns {Promise<Object>} Progress data
   */
  static async getProgressById(token, progressId) {
    const res = await fetch(`${API_URL}/api/v1/progress/${progressId}`, {
      headers: getAuthHeaders(token),
    });
    if (!res.ok) throw new Error('Failed to fetch progress');
    return res.json();
  }

  /**
   * Get all progress records for an enrollment
   * @param {string} token - Authentication token
   * @param {string} enrollmentId - The enrollment ID
   * @returns {Promise<Object>} { progress: Array } - List of progress records
   */
  static async getProgressByEnrollment(token, enrollmentId) {
    const res = await fetch(`${API_URL}/api/v1/progress/enrollment/${enrollmentId}`, {
      headers: getAuthHeaders(token),
    });
    if (!res.ok) throw new Error('Failed to fetch enrollment progress');
    return res.json();
  }

  /**
   * Get user's progress summary for a specific course
   * @param {string} token - Authentication token
   * @param {string} courseId - The course ID
   * @returns {Promise<Object>} Progress summary data
   */
  static async getUserCourseProgressSummary(token, courseId) {
    const res = await fetch(`${API_URL}/api/v1/progress/course/${courseId}/summary`, {
      headers: getAuthHeaders(token),
    });
    if (!res.ok) throw new Error('Failed to fetch course progress summary');
    return res.json();
  }

  /**
   * Update progress
   * @param {string} token - Authentication token
   * @param {string} progressId - The progress ID
   * @param {Object} updateData - Data to update (progress, completedItems, etc.)
   * @returns {Promise<Object>} Updated progress data
   */
  static async updateProgress(token, progressId, updateData) {
    const res = await fetch(`${API_URL}/api/v1/progress/${progressId}`, {
      method: 'PATCH',
      headers: getAuthHeaders(token),
      body: JSON.stringify(updateData),
    });
    if (!res.ok) throw new Error('Failed to update progress');
    return res.json();
  }

  /**
   * Mark an item as completed
   * @param {string} token - Authentication token
   * @param {string} progressId - The progress ID
   * @param {string} itemId - The item ID to mark as completed
   * @returns {Promise<Object>} Updated progress data
   */
  static async markItemCompleted(token, progressId, itemId) {
    const res = await fetch(`${API_URL}/api/v1/progress/${progressId}/complete-item`, {
      method: 'POST',
      headers: getAuthHeaders(token),
      body: JSON.stringify({ itemId }),
    });
    if (!res.ok) throw new Error('Failed to mark item completed');
    return res.json();
  }

  /**
   * Get progress for a specific module (instructor view)
   * @param {string} token - Authentication token
   * @param {string} moduleType - The module type (course, quiz, lesson, chapter)
   * @param {string} moduleId - The module ID
   * @returns {Promise<Object>} { progress: Array } - List of progress records
   */
  static async getProgressByModule(token, moduleType, moduleId) {
    const res = await fetch(`${API_URL}/api/v1/progress/module/${moduleType}/${moduleId}`, {
      headers: getAuthHeaders(token),
    });
    if (!res.ok) throw new Error('Failed to fetch module progress');
    return res.json();
  }

  /**
   * Delete progress (admin only)
   * @param {string} token - Authentication token
   * @param {string} progressId - The progress ID
   * @returns {Promise<Object>} Deletion result
   */
  static async deleteProgress(token, progressId) {
    const res = await fetch(`${API_URL}/api/v1/progress/${progressId}`, {
      method: 'DELETE',
      headers: getAuthHeaders(token),
    });
    if (!res.ok) throw new Error('Failed to delete progress');
    return res.json();
  }
}

export default ProgressService;
